GPUSimulators/SHMEMSimulator.py

```diff
@@ -0,0 +1,252 @@
+            self.east = None
+            new_boundary_conditions.east = boundary_conditions.east;
+        if (gj == grid.grid[1]-1 and boundary_conditions.north != Simulator.BoundaryCondition.Type.Periodic):
+            self.north = None
+            new_boundary_conditions.north = boundary_conditions.north;
+        sim.setBoundaryConditions(new_boundary_conditions)
+                
+        #Get number of variables
+        self.nvars = len(self.getOutput().gpu_variables)
+        
+        #Shorthands for computing extents and sizes
+        gc_x = int(self.sim.getOutput()[0].x_halo)
+        gc_y = int(self.sim.getOutput()[0].y_halo)
+        nx = int(self.sim.nx)
+        ny = int(self.sim.ny)
+        
+        #Set regions for ghost cells to read from
+        #These have the format [x0, y0, width, height]
+        self.read_e = np.array([  nx,    0, gc_x, ny + 2*gc_y])
+        self.read_w = np.array([gc_x,    0, gc_x, ny + 2*gc_y])
+        self.read_n = np.array([gc_x,   ny,   nx,        gc_y])
+        self.read_s = np.array([gc_x, gc_y,   nx,        gc_y])
+        
+        #Set regions for ghost cells to write to
+        self.write_e = self.read_e + np.array([gc_x, 0, 0, 0])
+        self.write_w = self.read_w - np.array([gc_x, 0, 0, 0])
+        self.write_n = self.read_n + np.array([0, gc_y, 0, 0])
+        self.write_s = self.read_s - np.array([0, gc_y, 0, 0])
+        
+        #Allocate data for receiving
+        #Note that east and west also transfer ghost cells
+        #whilst north/south only transfer internal cells
+        #Reuses the width/height defined in the read-extets above
+        self.in_e = np.empty((self.nvars, self.read_e[3], self.read_e[2]), dtype=np.float32)
+        self.in_w = np.empty((self.nvars, self.read_w[3], self.read_w[2]), dtype=np.float32)
+        self.in_n = np.empty((self.nvars, self.read_n[3], self.read_n[2]), dtype=np.float32)
+        self.in_s = np.empty((self.nvars, self.read_s[3], self.read_s[2]), dtype=np.float32)
+        
+        #Allocate data for sending
+        self.out_e = np.empty_like(self.in_e)
+        self.out_w = np.empty_like(self.in_w)
+        self.out_n = np.empty_like(self.in_n)
+        self.out_s = np.empty_like(self.in_s)
+        
+        self.logger.debug("Simlator subdomain {:d} initialized on context {:s}".format(self.index, sim.context))
+    
+        
+    def substep(self, dt, step_number):
+        self.exchange()
+        self.sim.substep(dt, step_number)
+    
+    def getOutput(self):
+        return self.sim.getOutput()
+        
+    def synchronize(self):
+        self.sim.synchronize()
+        
+    def check(self):
+        return self.sim.check()
+    
+    def computeDt(self):
+        local_dt = np.array([np.float32(self.sim.computeDt())])
+        global_dt = np.empty(1, dtype=np.float32)
+        self.grid.comm.Allreduce(local_dt, global_dt, op=MPI.MIN)
```

Comment:
MPI use here. Should be changed to a for loop to gather the minimum dt across the simulators